tool/site/client/src/components/sessionTimer.tsx

```diff
@@ -14,61 +15,77 @@ export const SessionTimer: React.FC = () => {
   const {baseUrl} = useContext(AppContext)
   const [cookies] = useCookies()
 
-  const lastTokenFetched = cookies["Last-Token-Fetched"]
+  const accessTokenFetched = cookies["Access-Token-Fetched"]
 
-  const calculateTimeLeft = () => {
+  const calculateTimeToTokenExpiry = () => {
     const now = getUtcEpochSeconds()
-    const justLessThenTenMinutes = 597
-    const difference = justLessThenTenMinutes - (now - lastTokenFetched)
-    let timeLeft = {}
+    return cookies["Token-Expires-In"] - (now - accessTokenFetched)
+  }
 
-    if (difference > 0) {
-      timeLeft = {
-        minutes: Math.floor((difference / 60) % 60),
-        seconds: Math.floor((difference) % 60)
-      }
-    }
+  const [tokenExpiresIn, setTokenExpiresIn] = useState(calculateTimeToTokenExpiry())
+  const [nextRefreshTime, setNextRefreshTime] = useState(cookies["Next-Refresh-Time"])
 
-    return timeLeft
+  const refreshToken = async() => {
+    const result = (await axiosInstance.post(`${baseUrl}auth/refresh`)).data
+    if (isRedirect(result)) {
+      redirect(result.redirectUri)
+      return {nextRefreshTime: 9999999999999}
+    }
+    return result
   }
 
-  const [timeLeft, setTimeLeft] = useState(calculateTimeLeft())
-
   const nonRedirectRoutes = [`${baseUrl}login`, `${baseUrl}logout`, `${baseUrl}prescribe/send`]
   const [redirectRequired, setRedirectRequired] = useState(
     !nonRedirectRoutes.includes(window.location.pathname)
   )
 
   useEffect(() => {
     setTimeout(() => {
-      setTimeLeft(calculateTimeLeft())
+      if (tokenExpiresIn <= 0) {
+        if (redirectRequired) {
+          setRedirectRequired(false)
+          redirect(`${baseUrl}logout`)
+        }
+      }
+      setTokenExpiresIn(calculateTimeToTokenExpiry())
     }, 1000)
   })
 
-  const timerIntervals = []
+  useEffect(() => {
+    const refreshRequired = getUtcEpochSeconds() > nextRefreshTime
+    if (refreshRequired) {
+      refreshToken().then(result => setNextRefreshTime(result.nextRefreshTime))
+    }
+  })
```

Comment:
split of these two useEffects has me confused, could we have the logic in this lower one run first in the top useEffect? looks like the dependency array is empty for both so the both just run once on render?

Reply:
consolidated